File: Concurrente/counter/counter.py
#! /usr/bin/env python

import threading

THREADS = 2
MAX_COUNT = 10000000
n = 0

def count():
    global n
    print("Thread {}".format(threading.current_thread().name))
    for i in range(MAX_COUNT//THREADS):
        n += 1
    print(f"El procés {threading.current_thread().name} ha comptat fins a {n}")
def main():
    threads = []
    for i in range(THREADS):
        # Create new threads
        t = threading.Thread(target=count)
        threads.append(t)
        t.start() # start the thread
    # Wait for all threads to complete
    for t in threads:
        t.join()
    print("Counter value: {} Expected: {}\n".format(n, MAX_COUNT))

if __name__ == "__main__":
    main()

File: Concurrente/counter/Counter.java
package counter;

public class Counter implements Runnable {
//public class Counter extends Thread {
    static final int THREADS = 2;
    static final int MAX_COUNT = 10000000;
    static volatile int n = 0;
//    static  int n = 0; // El compilador manté una còpia local a cada fil
    int id;

    public Counter(int id) {
        this.id = id;
    }

    @Override
    public void run() {
        int max = MAX_COUNT/THREADS;
        System.out.printf("Thread %d\n", id);
        for (int i =0; i < max; i++) {
            n += 1;
        }
    }

    public static void main(String[] args) throws InterruptedException {
        Thread[] threads = new Thread[THREADS];
        int i;

        for (i=0; i< THREADS; i++) {
            threads[i] = new Thread(new Counter(i));
            threads[i].start();
        }

        for (i=0; i< THREADS; i++) {
            threads[i].join();
        }
        System.out.printf("Counter value: %d Expected: %d\n", n, MAX_COUNT);
    }
}
